nextstrain_profiles/nextstrain-open/auspice_configuration.smk

```diff
@@ -0,0 +1,93 @@
+    size = build_info['build_size']['n']
+    if pretty_build_name == "Global":
+        build_info["title"] = f"Genomic epidemiology of novel coronavirus targetting {size} Global sequences"
+    else:
+        build_info["title"] = f"Genomic epidemiology of novel coronavirus focusing on {pretty_build_name} & targetting {size} sequences"
+    build_info['auspice_config'] = f"results/{build_name}/auspice_config.json"
+
+base_config = {
+    "title": "This title is not used (but is needed for validation)",
+    "build_url": "https://github.com/nextstrain/ncov",
+    "maintainers": [
+        {"name": "the Nextstrain team", "url": "https://nextstrain.org/"}
+    ],
+    "data_provenance": [
+        {"name": "GENBANK", "url": "https://www.ncbi.nlm.nih.gov/genbank/"}
+    ],
+    "colorings": [
+        { "key": "emerging_lineage", "title": "Emerging Lineage", "type": "categorical" },
+        { "key": "clade_membership", "title": "Clade", "type": "categorical"},
+        { "key": "S1_mutations", "title": "S1 mutations", "type": "continuous"},
+        { "key": "logistic_growth", "title": "Logistic growth", "type": "continuous"},
+        { "key": "current_frequency", "title": "Current frequency", "type": "continuous"},
+        { "key": "division", "title": "Admin Division", "type": "categorical" },
+        { "key": "country", "title": "Country", "type": "categorical" },
+        { "key": "region", "title": "Region", "type": "categorical" },
+        { "key": "host", "title": "Host", "type": "categorical" },
+        { "key": "author", "title": "Authors", "type": "categorical" },
+        { "key": "recency", "title": "Submission Date", "type": "categorical" },
+        { "key": "genbank_accession", "type": "categorical" }
+    ],
+    "geo_resolutions": [
+        "division",
+        "country",
+        "region"
+    ],
+    "display_defaults": {
+        "color_by": "clade_membership",
+        "distance_measure": "num_date",
+        "geo_resolution": "SET_IN_RULE",
+        "map_triplicate": "SET_IN_RULE",
+        "branch_label": "clade",
+        "transmission_lines": False
+    },
+    "filters": [
+        "recency",
+        "emerging_lineage",
+        "clade_membership",
+        "region",
+        "country",
+        "division",
+        "host"
+    ],
+    "panels": [
+        "tree",
+        "map",
+        "entropy",
+        "frequencies"
+    ]
+}
+
+rule make_auspice_config:
+    message: "Making auspice config file {output.auspice_config}"
+    output:
+        auspice_config = "results/{build_name}/auspice_config.json"
+    run:
+        import json
+        from copy import deepcopy
+        config = deepcopy(base_config)
+        config["display_defaults"]["geo_resolution"] = "division" if (wildcards.build_name.startswith("north-america") or wildcards.build_name.startswith("oceania")) else "country"
+        config["display_defaults"]["map_triplicate"] = wildcards.build_name.startswith("global")
```

Comment:
This type of workflow configuration might be clearer and easier to discover and maintain if it lived in the builds configuration instead of Snakemake logic. For example, we can get nearly the same outcomes as this custom `.smk` file produces with a configuration file like this:

```yaml
# Define a default Auspice configuration for open builds.
files:
  auspice_config: nextstrain_profiles/nextstrain/open/auspice_config.json

# Define builds.
builds:
  global:
    subsampling_scheme: nextstrain_region_global
    # Override the default auspice config where map_triplicate=false.
    map_triplicate: true

  africa:
    subsampling_scheme: nextstrain_region
    region: Africa

  asia:
    subsampling_scheme: nextstrain_region
    region: Asia

  europe:
    subsampling_scheme: nextstrain_region
    region: Europe

  north-america:
    subsampling_scheme: nextstrain_region
    region: North America
    # Override the default Auspice config value of geo_resolution=country.
    geo_resolution: division

  oceania:
    subsampling_scheme: nextstrain_region
    region: Oceania
    # Override the default Auspice config value of geo_resolution=country.
    geo_resolution: division

  south-america:
    subsampling_scheme: nextstrain_region
    region: South America
```

In this approach, the auspice config JSON already exists as a valid config that will produce reasonable outputs for any builds and we don't need a custom rule to create it per build. We pass through arguments to `augur export v2` from the builds config to override the JSON from the command line. This approach moves the bulk of this `.smk` file into a JSON file and leaves the custom title building to be resolved.

Reply:
This would also work, and (like my solution) has the big advantage that we don't have a lot of nearly-identical auspice config files hanging around as we currently do for the `nextstrain` profile. If we go this way,  I would propose changing these keys to (e.g.) `default_geo_resolution` as `geo_resolution` is ambiguous. 

Generally, I find a per-profile snakemake file which contains all the auspice config logic (and includes comments) easier to comprehend than a per-profile auspice config JSON which is then modified by rules later on in ways that are hard to remember; however I can understand this is personal preference. 

One advantage of programmatically generating the config files is when we extend their contents to (e.g.) specify custom legend bounds for S1 mutations. This depends on intermediate files (e.g. what's the range of S1 mutations for this build?) which is harder to generate up-front in a config file.

Reply:
@jameshadfield and I just chatted about this. Here's a summary of our discussion:

  - Dynamically modifying `config["builds"]` and dynamically creating Auspice config JSONs are separate issues here. We primarily discussed the auspice configs. Whether we need multiple build sizes seems to depend on whether there is scientific value in creating these and then the technical solution is another separate issue.
  - We thought of different ways to define build-specific auspice config parameters that override some default
    1. Programmatically build a valid Auspice config JSON with a Snakemake rule that overrides values in a "base config" defined in the rule's scope (top-level of the Snakemake file)
    2. Expose command line arguments for `augur export v2` through the `builds` definitions (as in my examples above) and add any missing arguments to Augur as needed (like `--default-map-triplicate`)
    3. Implement config file overlays in `augur export` such that `--auspice-config` can accept multiple inputs and deep merge these files (analogous to how Snakemake merges multiple inputs to `--configfile`). Define a base config JSON with most of the contents and minimal build-specific JSONs (e.g., defining only `map_triplicates: true`) to be merged with the base config.
    4. Implement config file overlays in the Snakemake workflow with a conditional rule that merges multiple JSONs in the order they are defined. The script we write to handle this logic and the config interface paves the way for migrating that functionality to `augur export`.
  - We agreed that a single concrete auspice config JSON is easier to inspect and understand than a combination of a base auspice config and many overriding command line arguments to `augur export`.
  - We also agreed that our workflow would be more robust if its own configuration (the `builds.yaml`) wasn't bound tightly to Auspice specifics. For example, exposing a `map_triplicate` parameter in the builds config through an Augur command line argument creates a tight link from the workflow to Augur to Auspice. In contrast, if we specify Auspice settings through Auspice config JSONs, the workflow and Augur only need to know about those files.
  - Dynamically generating Auspice legends bounds for S1 mutations, for example, is an interesting case where the implementation requires domain knowledge that Auspice does not have and effectively promotes the "config" to "data" by producing a data-dependent configuration. This might be unavoidably complex, so the question is where to put the complexity.

Reply:
Following conversations on Slack, 212b0d3 moves us to a simpler implementation of one auspice config per build. This discussion can shift over to https://github.com/nextstrain/ncov/issues/661 to allow this PR to be merged😄